### ISSUE TITLE
Reversed punctuation in RTL subtitles in Kodi 20 RC2

### ISSUE THREAD

**Issue by purpleman2k** [xbmc/xbmc#22398]
## Bug report
Punctuation is reversed when using Hebrew subtitles in Kodi Nexus RC2.
Likely applies to all other RTL languages too.

### Describe the bug
Punctuation should be placed at the end of the sentence, see attached screenshot for reference.

## Expected Behavior
Punctuation should be placed at the end of the sentence, see attached screenshot for reference.

## Actual Behavior
Punctuation is placed in the beginning of the sentence.

### To Reproduce

1. Change subtitles language in Player settings to Hebrew, character set to Hebrew (Windows).
2. Place attached subtitle file [Hebrew_subs.zip](https://github.com/xbmc/xbmc/files/10348200/Hebrew_subs.zip) in same dir of a video named accordingly (e.g. "Hebrew_subs.mp4").
3. Play the video in Kodi

### Debuglog

The debuglog can be found here:

### Screenshots 
Here are some links or screenshots to help explain the problem:

![image](https://user-images.githubusercontent.com/8142572/210668791-c0e1df90-8c1f-446e-9a41-6e5cca03a861.png)

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

This is the exact same issue that was fixed in a previous version of Kodi:
https://github.com/xbmc/xbmc/issues/18648

I copied the screenshots and supporting files from that issue because the symptom is the exact same and it reproduces with those files.

### Your Environment
Used Operating system:

 - [X ] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] macOS
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name:
 - Kodi version: 20 RC2

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by CastagnaIT** [comment]
log with debug must be provided
and if you are using a custom font please attach also the font file used

**Comment by scott967** [comment]
Looking at the sample text, I think there is an ambiguity that can cause this issue.

The first line of the provided text is
3F F9 EE F2 FA

That, is the "?" is the first byte.

I believe that the proper way of entering text is in logical order, which would be
F9 EE F2 FA 3F

But even in this byte sequence it should be treated as
3F   FA F2 EE F9
|->  <------------------------------
that is left-justified.

If the second byte ordering was used I think it would be 
                     3F FA F2 EE F9
<----------------------------|

**Comment by CastagnaIT** [comment]
@scott967 i have another hebrew srt you provided me a lot time ago
[Forum_Scott_RTL_hebrew.srt.txt](https://github.com/xbmc/xbmc/files/10404111/Forum_Scott_RTL_hebrew.srt.txt)
for you this looks correctly?
there is the line:
```
9
00:01:58,000 --> 00:02:01,450
```
that compared with MPV player looks different

on commit https://github.com/xbmc/xbmc/pull/20458/commits/d5726e21cbb3bf26d60551578182c274529436e2 i have fixed RTL case at least on webvtt
i think that if you remove this should become displayed on right way, but for sure will break webvtt
a sample: [shintel_s1e1_hebrew.vtt.txt](https://github.com/xbmc/xbmc/files/10404167/shintel_s1e1_hebrew.vtt.txt) that make use of `&rlm;`

**Comment by purpleman2k** [comment]
@CastagnaIT

Regarding the fonts, it happens on any font you use, including the default stock one.
The subtitle file you provided above plays fine in VLC and MPC-BE, but has the same issue I described when played with Kodi -- punctuations are appearing at the beginning of the sentence instead of the end.

For some more context, it is common practice with all RTL subtitles (and specifically Hebrew) the punctuation marks are technically placed within the SRT file at the beginning of the sentence (even though they are intended to be displayed at the end of the sentence). This is a **compensation** for misplacement with virtually every player out there -- but it's a necessary one, and I'll explain why.

You could argue that the error is in the SRTs themselves (because of the abovementioned practice of compensation in the authoring process, and that Kodi is actually doing it right and displaying as-is. Therefore, just fix the SRTs and there's nothing wrong with Kodi. So... not really. There's a reason why all other players haven't "fixed" that in the same way and just forced people to stop compensating in the SRTs. It appears that on most platforms, if you author the SRT as it's meant to be displayed and the player 'correctly' displays it as RTL -- like Kodi does now, it fixes some of the punctuation problems, but introduces other issues such as numeric placements that can't even be compensated for in the SRT -- any combination would display incorrectly. For example, a sentence that starts with a number would display the number at the end of the sentence instead of the beginning.

So bottom line, Kodi should revert back to how it used to interpret these subtitles, and how they are correctly read and displayed on other players :)

Hope this helps!

**Comment by BladeTrinity** [comment]
Again, **punctuation is reversed also** when using Hebrew subtitles in the new release which is **Kodi vertion 20.00** 
Please fix it.

**Comment by barronen1** [comment]
I’m experiencing the same issue on Kodi 20.

**Comment by guandms** [comment]
This also happens to me, in Kodi 19 everything was working fine, I tried copying the Fonts but the issue remains

**Comment by Pharaoh2k** [comment]
@CastagnaIT RTL punctuation was working fine on all previous versions until my shield auto-updated to v20.00 release.
Is it possible to get a test build without this https://github.com/xbmc/xbmc/commit/d5726e21cbb3bf26d60551578182c274529436e2 commit to confirm if this indeed corrects the issue?

**Comment by scott967** [comment]
I am back with my dev system.  But in reviewing things the root issue is that the unicode bidi algorithm requires that a text stream is first segmented into "paragraphs" and each paragraph must be assigned a base direction.  Unfortunately existing  subtitle specs (looked at subrip and advanced substa alpha) don't provide for setting the base direction.  So subtitle players typically just default to LTR as base.  The bidi algorithm can work with this, but neutral characters like punctuation will display correctly only when embedded in a text run.  Leading and trailing neutral chars always use the base direction.  I looked at various subtitle editors like Aegis-sub, and they provide work-arounds by reversing the punctuation in the sub files.  This results in visually rendered RTL text looking correct (but the default placement would be left-justified).  This work-around is technically WRONG.  The correct solution would be to use the defined unicode codepoints that signal RTL in the subtitle text and properly store the text.  Unfortunately we have to accept the subtitle editing environment the way it is, not the way it should be.

**Comment by Pharaoh2k** [comment]
@scott967 Yes, as you pointed out - the implementation is wrong, but that's been the industry standard for a while and worked well on all previous Kodi releases. If it ain't broken...

**Comment by scott967** [comment]
The entire subtitle render system was effectively broken and has been redone for 20, but the issue IMO is valid and needs a fix.

**Comment by maan-doar** [comment]
Perhaps a compromise solution would be to add another "pseudo encoding" to the list of possible subtitle encodings, in Kodi settings.
Something like _"Hebrew (Windows, LTR)"_ - which internally would translate to codpepage 1255, with **base direction LTR**.
This would be compatible to pre-20 implementation of _"Hebrew (Windows)"_, and would work with existing srt's.
The encoding would be _"Hebrew (Windows, RTL)"_ - which is codepage 1255, with **base direction RTL**.
I think I saw this once in another video player, or perhaps it was a dedicated streamer.

**Comment by maan-doar** [comment]
> Unfortunately existing subtitle specs (looked at subrip and advanced substa alpha) don't provide for setting the base direction. So subtitle players typically just default to LTR as base. 

@scott967 , in fact, when the player assumes LTR as base, (and passes this to fribid), the existing de-facto "standard" of subtitles display **correctly**

**Comment by barronen1** [comment]
@CastagnaIT hi any progress?

**Comment by CastagnaIT** [comment]
yes partial revert works but break webvtt decoder
is needed to do a change in webvtt decoder and i am not sure about these changes

i have an arabic subtitle to test, i need someone able to read arabic language, someone here?

**Comment by barronen1** [comment]
> yes partial revert works but break webvtt decoder is needed to do a change in webvtt decoder and i am not sure about these changes
> 
> i have an arabic subtitle to test, i need someone able to read arabic language, someone here?

How does it work in Kodi 19 that RTL is correct?

**Comment by Pharaoh2k** [comment]
> yes partial revert works but break webvtt decoder is needed to do a change in webvtt decoder and i am not sure about these changes
> 
> i have an arabic subtitle to test, i need someone able to read arabic language, someone here?

I can read both Hebrew and Arabic, both are RTL and suppose to work the same way.
How may I help?

**Comment by CastagnaIT** [comment]
@Pharaoh2k i have already tested hebrew that is ok

but i need a feedback for webvtt use case that make use of `&lrm;` (left-to-right)
from what i understand when &lrm; is used should display text in a different order

play from Website service:
![immagine](https://user-images.githubusercontent.com/3257156/215169976-e491e8c7-84f4-4e2b-8ea1-39233a15ef40.png)

play from Kodi:
![immagine](https://user-images.githubusercontent.com/3257156/215170452-0daff92c-37ea-47fb-a551-c2c27e55ee2e.png)

subtitle file:
[LRM_Wet.Hot.American.Summer.First.Day.of.Camp.S01E01.Campers.Arrive.vtt.txt](https://github.com/xbmc/xbmc/files/10522249/LRM_Wet.Hot.American.Summer.First.Day.of.Camp.S01E01.Campers.Arrive.vtt.txt)
position 00:02:15.927 ID 30

What can you tell me?

**Comment by scott967** [comment]
> 
> How does it work in Kodi 19 that RTL is correct?

It isn't correct in Kodi 19; just an accident that it works in your use-case.

**Comment by scott967** [comment]
Char set potential issues:
These I think are the char set encodings that need to be considered:
ISO-8859-6 Arabic
ISO 8859-8 Hebrew with additional complications of possible LRM/RLM and LRE/RLE extensions
UTF-8

IIUC there was an assumption made that if char set is 8859-6 or 8859-8 that base direction should be set to RTL but due to VSfilter? limitations fan-subs in these char sets assume base dir of LTR.

I don't know if SSA/ASS has a style that is commonly used to set base dir RTL

WebVTT does mandate the setting of base dir so I think we should obey the spec and handle accordingly.

**Comment by maan-doar** [comment]
> How does it work in Kodi 19 that RTL is correct?
Kodi 19 (and before) assumed LTR as base direction (yes, Left-To-Right) for Hebrew subtitles not in unicode (most commonly CP1255, or ISO-8859-8),
and this matched the **convention** that the subtitle file creators followed.
The key word is "assumed", since, as mentioned here above, the .srt format does not have a notion of declaring "base direction".
(and I think neither does ASS/SSA)

**Comment by maan-doar** [comment]
> IIUC there was an assumption made that if char set is 8859-6 or 8859-8 that base direction should be set to RTL but due to VSfilter? limitations fan-subs in these char sets assume base dir of LTR.

More because of a limitation of authoring tools and how they display the preview of subs on video. Yes, SubtitleWorkhop, I am looking at you ;-)

**Comment by CastagnaIT** [comment]
on ASS format specs there is the "Encoding" field that can be used to specify the type encoding to apply, as number

but libass library allow/support only two "Encoding" ways, are:
`-1 == FRIBIDI_PAR_ON`
`any other number == fall always to: FRIBIDI_PAR_LTR`
(fribidi definitions: https://manpages.debian.org/unstable/libfribidi-dev/fribidi_get_par_embedding_levels.3.en.html#PARAMETERS)
ref. https://github.com/xbmc/xbmc/commit/d5726e21cbb3bf26d60551578182c274529436e2#diff-78b3eb5f8ef02a0e30d34608f683579549ae1f5f479d1bb4728abbe60ace815dR326-R328

so usually FRIBIDI_PAR_LTR is "always" set,
this looks like a problem only for webvtt cases that use `&lrm;`, `&rlm;` optionals unicode marks escapes

> WebVTT does mandate the setting of base dir so I think we should obey the spec and handle accordingly

Unfortunately i noticed that webvtt is also not working properly on Kodi 20
currently is displayed correctly only when `&lrm;`, `&rlm;` are set on each subtitle text line
since are optionals, you can have a webvtt file without `&rlm;` and is currently displayed wrong

so if we revert the libass change
https://github.com/xbmc/xbmc/commit/d5726e21cbb3bf26d60551578182c274529436e2#diff-78b3eb5f8ef02a0e30d34608f683579549ae1f5f479d1bb4728abbe60ace815dR326-R328
SRT and webvtt without `&lrm;`, `&rlm;`, are displayed correctly
but webvtt with `&lrm;`, `&rlm;`, are not displayed correctly (as screenshots on #20458)

I have fear that without going out-of-spec we cannot solve

**Comment by Barlerer** [comment]
I can confirm downgrading from version 20.0 to 19.5 has solved the issue for me as a temporary solution!
(Hebrew)

**Comment by psp2111-ADSLGATE** [comment]
> @Pharaoh2k i have already tested hebrew that is ok
> 
> but i need a feedback for webvtt use case that make use of `&lrm;` (left-to-right) from what i understand when ‎ is used should display text in a different order
> 
> play from Website service: ![immagine](https://user-images.githubusercontent.com/3257156/215169976-e491e8c7-84f4-4e2b-8ea1-39233a15ef40.png)
> 
> play from Kodi (with fixes): ![immagine](https://user-images.githubusercontent.com/3257156/215170452-0daff92c-37ea-47fb-a551-c2c27e55ee2e.png)
> 
> subtitle file: [LRM_Wet.Hot.American.Summer.First.Day.of.Camp.S01E01.Campers.Arrive.vtt.txt](https://github.com/xbmc/xbmc/files/10522249/LRM_Wet.Hot.American.Summer.First.Day.of.Camp.S01E01.Campers.Arrive.vtt.txt) position 00:02:15.927 ID 30
> 
> What can you tell me?

The second pic with "play from Kodi (with fixes)" is the correct order and makes sense when reading it

**Comment by barronen1** [comment]
Hi, any progress with that?

It would really help us if it get fixed in next Kodi 20.1 release, atleast for Hebrew subtitles

**Comment by BladeTrinity** [comment]
Hi
I also agree and joining to the request of **barronen1**.
Please fix it in the next release.
Thank you.

**Comment by barronen1** [comment]
What does it mean?

**Comment by Pharaoh2k** [comment]
> What does it mean? @CastagnaIT

It's a pull request with the fix details to be added to Kodi:
https://github.com/xbmc/xbmc/pull/22663/commits/be401a3ab0c59cc2a3308300a02d230982cab11d

**Comment by barronen1** [comment]
> > What does it mean? @CastagnaIT
> 
> It's a pull request with the fix details to be added to Kodi: [be401a3](https://github.com/xbmc/xbmc/commit/be401a3ab0c59cc2a3308300a02d230982cab11d)

I see. I saw the now the merge request but with failed checks :(

**Comment by CastagnaIT** [comment]
for those who keep complaining because they are in a hurry, compile Kodi yourself with the PR fix so youre happy
if not so, rollback to Kodi 19 or wait, because the next Kodi release will be in a few months, and the fix PR can wait for further developments

**Comment by barronen1** [comment]
You mean 20.1 public release generally will be at least in a few months?

And how do I recompile Kodi 20.0 APK with the PR fix?

**Comment by Pharaoh2k** [comment]
> for those who keep complaining because they are in a hurry, compile Kodi yourself with the PR fix so youre happy if not so, rollback to Kodi 19 or wait, because the next Kodi release will be in a few months, and the fix PR can wait for further developments

If I am not mistaken, the milestone set on the PR is Kodi 21, so he may have meant Kodi 21, not 20.1.
Hopefully, @CastagnaIT can clarify.

**Comment by CastagnaIT** [comment]
backports will be done only after the main PR will be approved and not before,
nothing prevent you to cherry-picking the commit itself and build over previous versions

**Comment by Pharaoh2k** [comment]
> backports will be done only after the main PR will be approved and not before, nothing prevent you to cherry-picking the commit itself and build over previous versions

Yes, sure, you have already pointed that option out. 
We are all just wondering what you meant by "the next Kodi release will be in a few months". Did you mean v21 or v20.x?

**Comment by CastagnaIT** [comment]
> Did you mean v21 or v20.x?

I thought it was implied talking about v20, you cannot think that v21 will be released in a couple of months...🙂

**Comment by Pharaoh2k** [comment]
> > Did you mean v21 or v20.x?
> 
> I thought it was implied talking about v20, you cannot think that v21 will be released in a couple of months...🙂

lol...a few months can mean different things to different people....now that you say a "couple" of months that's clearer ;)

**Comment by CastagnaIT** [comment]
i dont want talk about ETA, therefore it is only an indicative time, not full established
the release date may change according to the needs of the team and development requirements

**Comment by barronen1** [comment]
> backports will be done only after the main PR will be approved and not before,
> nothing prevent you to cherry-picking the commit itself and build over previous versions

Is there a tutorial for building kodi apk over the source code from github? I’ll use your commit with the fixed RTL

**Comment by barronen1** [comment]
> > > backports will be done only after the main PR will be approved and not before,
> > > nothing prevent you to cherry-picking the commit itself and build over previous versions
> > 
> > 
> > Is there a tutorial for building kodi apk over the source code from github? I’ll use your commit with the fixed RTL
> 
> Not sure if it's up to date but you may wanna check it out: https://github.com/xbmc/xbmc/blob/master/docs/README.Android.md

I’ve read it and I am afraid it’s too complicated for me. What are the chances anyone can Build Kodi 20.0 APK with the fixed RTL commit for arm (32bit) and aarch64 (64bit), and upload it?

**Comment by barronen1** [comment]
> > > backports will be done only after the main PR will be approved and not before,
> > > nothing prevent you to cherry-picking the commit itself and build over previous versions
> > 
> > 
> > Is there a tutorial for building kodi apk over the source code from github? I’ll use your commit with the fixed RTL
> 
> Not sure if it's up to date but you may wanna check it out: https://github.com/xbmc/xbmc/blob/master/docs/README.Android.md

The guide is up to date except one mistake in one line which can fail a lot of copy-paste users..

**Comment by fuzzard** [comment]
Feel free to PR an update to the docs if you have come across an issue. Systems evolve, and docs are "accurate" at a given point in time, but can be updated to match changes required.

**Comment by Pharaoh2k** [comment]
> > > > backports will be done only after the main PR will be approved and not before,
> > > > nothing prevent you to cherry-picking the commit itself and build over previous versions
> > > 
> > > 
> > > Is there a tutorial for building kodi apk over the source code from github? I’ll use your commit with the fixed RTL
> > 
> > 
> > Not sure if it's up to date but you may wanna check it out: https://github.com/xbmc/xbmc/blob/master/docs/README.Android.md
> 
> The guide is up to date except one mistake in one line which can fail a lot of copy-paste users..

What's the mistake? Did you manage to build the apk?

**Comment by barronen1** [comment]
@CastagnaIT 
I've managed to compile the APK from your subtitle_rtl_fix branch:
https://github.com/CastagnaIT/xbmc/tree/subtitle_rtl_fix

But the Kodi version is 21, and master branch is also version 21. How can I choose to compile the Kodi 20.0 latest public release? (And before compiling - edit manually the two files you edited to fix the RTL issue)

**Comment by barronen1** [comment]
> > > > > backports will be done only after the main PR will be approved and not before,
> > > > > nothing prevent you to cherry-picking the commit itself and build over previous versions
> > > > 
> > > > 
> > > > Is there a tutorial for building kodi apk over the source code from github? I’ll use your commit with the fixed RTL
> > > 
> > > 
> > > Not sure if it's up to date but you may wanna check it out: https://github.com/xbmc/xbmc/blob/master/docs/README.Android.md
> > 
> > 
> > The guide is up to date except one mistake in one line which can fail a lot of copy-paste users..
> 
> What's the mistake? Did you manage to build the apk?

I successfully compiled Kodi 20.0 + Hebrew RTL fix APKs (32/64 bit versions)
Using Kodi 20.0 Nexus branch + @CastagnaIT RTL fix code manually edited in the relevant files. 

And it does work now correctly for Hebrew subtitles.

**Comment by Pharaoh2k** [comment]
> > > > > > backports will be done only after the main PR will be approved and not before,
> > > > > > nothing prevent you to cherry-picking the commit itself and build over previous versions
> > > > > 
> > > > > 
> > > > > Is there a tutorial for building kodi apk over the source code from github? I’ll use your commit with the fixed RTL
> > > > 
> > > > 
> > > > Not sure if it's up to date but you may wanna check it out: https://github.com/xbmc/xbmc/blob/master/docs/README.Android.md
> > > 
> > > 
> > > The guide is up to date except one mistake in one line which can fail a lot of copy-paste users..
> > 
> > 
> > What's the mistake? Did you manage to build the apk?
> 
> I’ve successfully compiled Kodi 20.0 + Hebrew RTL fix APKs (32/64 bit versions) Using Kodi 20.0 Nexus branch + @CastagnaIT RTL fix code manually edited in the relevant files.
> 
> And it does work now correctly for Hebrew subtitles.

Excellent! 👍🏼
Do you mind sharing the apk?

**Comment by muttley79** [comment]
> > > > > > > backports will be done only after the main PR will be approved and not before,
> > > > > > > nothing prevent you to cherry-picking the commit itself and build over previous versions
> > > > > > 
> > > > > > 
> > > > > > Is there a tutorial for building kodi apk over the source code from github? I’ll use your commit with the fixed RTL
> > > > > 
> > > > > 
> > > > > Not sure if it's up to date but you may wanna check it out: https://github.com/xbmc/xbmc/blob/master/docs/README.Android.md
> > > > 
> > > > 
> > > > The guide is up to date except one mistake in one line which can fail a lot of copy-paste users..
> > > 
> > > 
> > > What's the mistake? Did you manage to build the apk?
> > 
> > 
> > I’ve successfully compiled Kodi 20.0 + Hebrew RTL fix APKs (32/64 bit versions) Using Kodi 20.0 Nexus branch + @CastagnaIT RTL fix code manually edited in the relevant files.
> > And it does work now correctly for Hebrew subtitles.
> 
> Excellent! 👍🏼 Do you mind sharing the apk?

I would also appreciate it if you could share the apks

**Comment by muttley79** [comment]
Sorry to nudge.. Any way we can get the apks?

**Comment by ntzb** [comment]
here's a build I did for Windows, today, with working RTL punctuation, thanks to the pull request from @CastagnaIT
https://www79.zippyshare.com/v/tLb1gupx/file.html
hopefully this will be merged soon so we'll have it in the official builds.

p.s. the installer might have missing/too many addons, I guess there are some differences between the official documentation for windows compilation, and the official builds available to download.

p.s.2 I tried, and failed in compiling for android, not sure what the problem might be.

**Comment by maan-doar** [comment]
> Sorry to nudge.. Any way we can get the apks?

You can get it from the test builds, for PR22663 [here](http://mirrors.kodi.tv/test-builds/android/arm/kodi-20230208-309bb197-PR22663-armeabi-v7a.apk)

**Comment by TwilightMercy** [comment]
> > Sorry to nudge.. Any way we can get the apks?
> 
> You can get it from the test builds, for PR22663 [here](http://mirrors.kodi.tv/test-builds/android/arm/kodi-20230208-309bb197-PR22663-armeabi-v7a.apk)

Where are all build tests apks?

**Comment by CastagnaIT** [comment]
i will post all builds in a few hours

**Comment by CastagnaIT** [comment]
Test builds:
Android64: http://mirrors.kodi.tv/test-builds/android/arm64-v8a/kodi-20230214-d8073163-subtitle_rtl_fix-arm64-v8a.apk
Android: http://mirrors.kodi.tv/test-builds/android/arm/kodi-20230214-d8073163-subtitle_rtl_fix-armeabi-v7a.apk
OSX64: http://mirrors.kodi.tv/test-builds/osx/x86_64/kodi-20230214-d8073163-subtitle_rtl_fix-28778-x86_64.dmg
Windows64: http://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20230214-d8073163-subtitle_rtl_fix-x64.exe

**Comment by Pharaoh2k** [comment]
[**@CastagnaIT**](https://github.com/CastagnaIT) commented on [Feb 14, 2023, 4:33 PM GMT+7](https://github.com/xbmc/xbmc/issues/22398#issuecomment-1429412739 "2023-02-14T09:33:08Z - Replied by Github Reply Comments"):
> Test builds:  
> AndroidArm64: [http://mirrors.kodi.tv/test-builds/android/arm64-v8a/kodi-20230214-d8073163-subtitle\_rtl\_fix-arm64-v8a.apk](http://mirrors.kodi.tv/test-builds/android/arm64-v8a/kodi-20230214-d8073163-subtitle_rtl_fix-arm64-v8a.apk)  
> AndroidArm: [http://mirrors.kodi.tv/test-builds/android/arm/kodi-20230214-d8073163-subtitle\_rtl\_fix-armeabi-v7a.apk](http://mirrors.kodi.tv/test-builds/android/arm/kodi-20230214-d8073163-subtitle_rtl_fix-armeabi-v7a.apk)  
> OSX64: [http://mirrors.kodi.tv/test-builds/osx/x86\_64/kodi-20230214-d8073163-subtitle\_rtl\_fix-28778-x86\_64.dmg](http://mirrors.kodi.tv/test-builds/osx/x86_64/kodi-20230214-d8073163-subtitle_rtl_fix-28778-x86_64.dmg)  
> Windows64: [http://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20230214-d8073163-subtitle\_rtl\_fix-x64.exe](http://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20230214-d8073163-subtitle_rtl_fix-x64.exe)

Nice. Hebrew subtitles are properly displayed now and even dolby vision seems to be fully supported.
Only one issue I noticed - can't adjust subtitles' position/height.

**Comment by SubEdict** [comment]
Exactly what I was afraid of. Now with @CastagnaIT fix, all the UTF-8 Hebrew that worked perfectly (embedded from services like Apple, Amazon, Netflix etc.) are now broken (reversed).

Embracing the "Half glass full", thanks for fixing the _external_ .srt 1255 subs, yes it works great as it was before v20.

**Comment by CastagnaIT** [comment]
@SubEdict on my tests confirmed by other user netflix show correctly RTL subtitles
so please provide a concrete example of not working RTL netflix subtitles by using test build above https://github.com/xbmc/xbmc/issues/22398#issuecomment-1429412739
i need to know also the title of movie or tvshow season/episode

**Comment by CastagnaIT** [comment]
> 90% of the cases

sentences of this kind dont help and are useless for investigating

> tested Apple and Amazon

i dont have that accounts, so to try investigate i need you attach here at least one original subtitle file (zipped or rename extension to txt) with a screenshot to understand the wrong/correct behaviour

**Comment by SubEdict** [comment]
Forget Netflix please, my bad.

> i need you attach here at least one original subtitle file

How? It's embedded in Apple stream, attached screenshot below:
https://i.imgur.com/mPxDQGP.png

**Comment by CastagnaIT** [comment]
are not embedded are differents streams, but depends from the type of stream

if we are luky one solution is attach a kodi log with debug enabled taken after played video, could be visible the downloadable webaddress file, or another way is open your browser of your video provider, analyze the http flow by using browser dev tools to find the subtitle stream then save it to a file

if you are not able to do it, suggest you to ping here the addons developers and ask them to help us

**Comment by SubEdict** [comment]
Thanks @CastagnaIT for yours suggestions. Will try and report if there's any progress.

**Comment by CastagnaIT** [comment]
hi @Varstahl, we need your help here for amaz@n use case,
im trying to fix subtitle RTL problem (e.g. Hebrew) where fixes can be tested with test builds on this comment https://github.com/xbmc/xbmc/issues/22398#issuecomment-1429412739
@SubEdict affirm that also using these fixes on Amazon the subtitles are still displayed wrong

now i have no idea if your addon make some kind of edit hack for subtitles i hope no
but to understand what happens i need you attach here the original subtitle stream file
can you provide original am@zon subtitle RTL hebrew file? so i can investigate what happens?

**Comment by Varstahl** [comment]
> now i have no idea if your addon make some kind of edit hack for subtitles i hope no

We do actually, in here: https://github.com/Sandmann79/xbmc/blob/43484d50b219dc416404783291420868b4dcf8d3/plugin.video.amazon-test/resources/lib/proxy.py#L373-L422

For starters we do DFXP/TTMLv2 to SRT conversions, then we fix a bunch of issues that cropped up over the years. For hebrew we don't really touch anything, but it's possible that something goes wrong in the conversion, although it might be that Amazon has issues with the marks/embeddings like we had fixed in the past for a different language.

I should be able to provide you with both the original and the transcoded subtitles, let me see if I can find something with hebrew subtitles.

**Comment by Varstahl** [comment]
@CastagnaIT here's the two files: [Hebrew Subtitles](https://github.com/xbmc/xbmc/files/10754174/he.zip). The `.ttml2` is as downloaded from their servers, the `.srt` is our conversion.

Let me know if there's anything I can do to help, or if I need to add more special cases for our subtitles parsing. Admittedly they do a pretty poor job with them overall, including but not limited to bad timestamps.

**Comment by CastagnaIT** [comment]
@SubEdict please specify what addon you use to play apple videos

@Varstahl yes not easy convert ttml to srt there are a lot of downsides,
however the ttml attached has not specified RTL property i think this suggest to use default LTR,
can you take a screenshot from website service of that video with subtitle at position 00:02:12 (shown below)
i want know if on the website is shown differently

at position 00:02:12 is shown on kodi with fixes (and also VLC) as follow:
![immagine](https://user-images.githubusercontent.com/3257156/219357815-42b304f0-ceea-4f91-a359-d56d626467e6.png)
someone can say if it is looks right?

**Comment by Pharaoh2k** [comment]
The Hebrew punctuation in the above brown screenshot looks wrong @CastagnaIT

**Comment by CastagnaIT** [comment]
ok then i will wait a website screenshot from @Varstahl to see if also on website is displayed wrong

**Comment by Varstahl** [comment]
![image](https://user-images.githubusercontent.com/284077/219362079-de9796d5-02d3-4693-87ba-8b209f714f0d.png)

**Comment by CastagnaIT** [comment]
however yes the situation is this
srt files to be displayed with RTL most of times apply (also by using editors) a reversed punctuation (as explained on PR #22663)
now you convert ttml to srt but dont have a reversed punctuation then is displayed wrong

i can suggest one of the following changes to your addon:
1) try apply the unicode emdedded direction as first character of each line (as per webvtt example PR #22663)
2) apply the reversed punctuation in similar way subtitle edit software do, should be this code https://github.com/SubtitleEdit/subtitleedit/blob/6e6f4744475b7502a6e64367c52f96cbcd843393/src/libse/Common/Utilities.cs#L1311-L1412

however yes it is not exactly a kodi problem, unfortunately RTL with SRT currently sucks...

**Comment by Varstahl** [comment]
> however yes it is not exactly a kodi problem

Yeah, no, Amazon sucks in this regard. There are many and more issues, with subtitles and audio tracks both. But it should be fixed in the addon itself, I'll open an issue for avod.

> i can suggest one of the following changes to your addon

I honestly would go back to TTMLv2, and then inject the RTL metadata, except for the fact that I still have to find where that metadata comes from, or will have to hard code everything, but again this goes in a separate non Kodi issue.

**Comment by scott967** [comment]
> 

Sorry I am behind on this as I do not have a source of subtitles other than random downloads from opensub.  I see in that code
`if filename.startswith("ar"):
                    from unicodedata import lookup
                    text = re.sub(r'^(?!{}|{})'.format(lookup('RIGHT-TO-LEFT MARK'), lookup('RIGHT-TO-LEFT EMBEDDING')),
                                  lookup('RIGHT-TO-LEFT EMBEDDING'), text, flags=re.MULTILINE)
                    text = text.replace('?', '؟').replace(',', '،')`

So I guess for Arabic if the sub does not start with RLM or RLE then RLE gets inserted which I guess works, though of course has no impact on Hebrew.  Is there some reason Arabic gets handled differently?

**Comment by scott967** [comment]
I looked at that he.ttml2 and the converted he.srt

As CastagnaIT suggests, the ttml2 source does not use markup to set direction, as is best practice per W3C.  The body of the subs is utf-8 which is good.  However, if they don't use markup the fallback is to use Unicode RLE code.  Same with the resulting srt file.  I opened the srt in subtitle edit 3.6.11, and that program lets me select all sub lines and in 2 clicks (Edit - Fix RTL via Unicode control character) it was fixed.  Of course, I "knew" that the subs were in Hebrew.  There's nothing in that ttml2 file to indicate language.  The only thing I can think of is to input the strings and see if the first letters are in unicode blocks for RTL scripts and then assume they are intended RTL.  In the exe CastagnaIT provided:

![hesub](https://user-images.githubusercontent.com/11218744/219923225-1dbe6347-9981-4376-aec6-69bb7fe57d26.jpg)

**Comment by Varstahl** [comment]
> So I guess for Arabic if the sub does not start with RLM or RLE then RLE gets inserted which I guess works, though of course has no impact on Hebrew. Is there some reason Arabic gets handled differently?

iirc hebrew used to work, while arabic only worked for some of the subs, hence the custom fix. We could apply the same to hebrew but I'll be looking at a more permanent fix rather than monkey patching it in.

> Of course, I "knew" that the subs were in Hebrew. There's nothing in that ttml2 file to indicate language.

The language is specified in the first row of the ttml (second, if we consider the XML header)

```xml
<tt xmlns="http://www.w3.org/ns/ttml" … xml:lang="he-IL">
```
